### PR TITLE
[FSDP] Fix `FSDP.clip_grad_norm_()` for `NO_SHARD`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#88955 [FSDP] Fix `FSDP.clip_grad_norm_()` for `NO_SHARD`**
* #88450 [FSDP] Introduce `ModuleWrapPolicy` for simplicity

This PR fixes `FSDP.clip_grad_norm_()` for `NO_SHARD`, which previously "double-counted" each gradient `world_size`-many times.

This does not address any discrepancies between `FULL_SHARD` and DDP. (Note that the unit tests do show parity between `FULL_SHARD` and DDP when using `FSDP.clip_grad_norm_()` and `nn.utils.clip_grad_norm_()` respectively on one iteration.)

The added unit test code path tests mixing nested FSDP instances with both `FULL_SHARD` and `NO_SHARD` to ensure that the `local_sharded_norm` and `local_nonsharded_norm` computations are interoperating correctly. I want to test non-FSDP root instance in the future, but this is BC breaking since we need to make `clip_grad_norm_()` a static method, which would require a different method call syntax (`FSDP.clip_grad_norm_(root_module, ...)` vs. `root_module.clip_grad_norm_(...)`).

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -1161,23 +1161,45 @@ def clip_grad_norm_(
             self._streams["unshard"],
             self._streams["pre_unshard"],
         )
-
```
Comment (zhaojuanmao) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1164]: For NO_SHARD, wondering whether we can just call the default PyTorch clip_grad_norm_() and early return without all_reduce and following logics?

**Review by zhaojuanmao — APPROVED** [review]

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1164]: Yes, we could, but this version enables us to mix and match `NO_SHARD` and `FULL_SHARD` for different submodules. If the user knows the entire FSDP instance is only `NO_SHARD`, then they can just use `torch.nn.utils.clip_grad_norm_()`.

I think some of this logic in `FSDP.clip_grad_norm_()` will be useful when thinking about how to write `clip_grad_norm_()` for our composable APIs.

Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1164]: We can optionally do a check over all FSDP instances in the module tree, and if all are `NO_SHARD`, then we can return `nn.utils.clip_grad_norm_()` directly as you suggested. Maybe I can include that fast path in a follow-up.

Comment (zhaojuanmao) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1164]: I see, that makes sense!
